iree/turbine/kernel/wave/layers/quant_linear.py

```diff
@@ -0,0 +1,293 @@
+    quant_params,
+    dynamic_dims: bool = False,
+    mfma_variant: MMAType = MMAType.F32_16x16x32_F8,
+    use_bias: bool = False,
+):
+    # Input sizes
+    B = tkl.sym.B
+    M = tkl.sym.M
+    N = tkl.sym.N
+    K = tkl.sym.K
+    # Workgroup tile sizes
+    BLOCK_B = tkl.sym.BLOCK_B
+    BLOCK_M = tkl.sym.BLOCK_M
+    BLOCK_N = tkl.sym.BLOCK_N
+    BLOCK_K = tkl.sym.BLOCK_K
+    # Address space (for GPU, shared(1) or global(0))
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+    # Other hyperparameters
+    LOAD_ELEMS_PER_THREAD = tkl.sym.LOAD_ELEMS_PER_THREAD
+    STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+    # Expose user-constraints
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(B, BLOCK_B, 2)]
+    constraints += [tkw.TilingConstraint(K, BLOCK_K)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M / 2)]
+    constraints += [tkw.WaveConstraint(N, BLOCK_N / 2)]
+
+    constraints += [
+        tkw.HardwareConstraint(
+            threads_per_wave=64,
+            waves_per_block=(2, 2, 1),
+            vector_shapes={B: 0},
+            mma_type=mfma_variant,
+        )
+    ]
+
+    # With dynamic dimensions, we need to add an assumption on how big
+    # the reduction dimension is to determine whether we can schedule or not.
+    if dynamic_dims:
+        constraints += [tkw.Assumption(K > BLOCK_K * 4)]
+
+    input_wtype = torch_dtype_to_wave(input_dtype)
+    [weight_scale, input_scale, quant_dtype] = quant_params
+    [qdtype_min, qdtype_max] = torch_dtype_range(quant_dtype)
+
+    def clamp_tensor(source_reg, lower_bound, upper_bound):
+        clamped = tkw.minimum(source_reg, upper_bound)
+        clamped = tkw.maximum(clamped, lower_bound)
+        clamped = tkw.cast(clamped, torch_dtype_to_wave(quant_dtype))
+        return clamped
+
+    # Wave-level micro-kernel.
+    # Since warps are not directly addressable, there is no
+    # explicit notion of a warp id (like a workgroup or thread id).
+    # This kernel uses the input sizes M, N, K throughout, as the tiling
+    # and data movement strategy is determined during the compilation process.
+    # These can be influenced by introducing constraints.
+    def gemm_core(a, b, c_reg, result):
+        # TODO: Registers for quantization scaling of inputs. Remove once scalar
+        # codegen is enabled.
+        a_scale = tkl.Register[B, M, K, input_wtype](1 / input_scale.item())
+        a_clamp_max = tkl.Register[B, M, K, input_wtype](qdtype_max)
+        a_clamp_min = tkl.Register[B, M, K, input_wtype](qdtype_min)
+        b_scale = tkl.Register[N, K, input_wtype](1 / weight_scale.item())
+        b_clamp_max = tkl.Register[N, K, input_wtype](qdtype_max)
+        b_clamp_min = tkl.Register[N, K, input_wtype](qdtype_min)
+        a_scale_deq = tkl.Register[B, M, N, input_wtype](input_scale.item())
```

Comment:
any thoughts on fusing this into a single `ab_scale_deq = tkl.Register[B, M, N, input_wtype](input_scale.item() * weight_sclae.item())` ?